core/src/apps/bitshares/base/operations.py

```diff
@@ -0,0 +1,215 @@
+from trezor import wire
+
+# fmt:off
+from trezor.messages.BitSharesOperation import BitSharesOperation
+from trezor.messages.BitSharesTransferOperation import BitSharesTransferOperation
+from trezor.messages.BitSharesLimitOrderCreateOperation import BitSharesLimitOrderCreateOperation
+from trezor.messages.BitSharesLimitOrderCancelOperation import BitSharesLimitOrderCancelOperation
+from trezor.messages.BitSharesAccountCreateOperation import BitSharesAccountCreateOperation
+from trezor.messages.BitSharesAccountUpdateOperation import BitSharesAccountUpdateOperation
+# fmt:on
+
+from .types import (
+    Bool,
+    Uint16,
+    Uint32,
+    String,
+    Permission,
+    OperationId,
+    ObjectId,
+    Optional,
+    Asset,
+    Memo,
+    EmptyExtension,
+    AccountOptions,
+)
+from .parsing import Node
+
+from ucollections import OrderedDict
+
+# Union type for all operations
+class Operation(Node):
+    def __init__(self, msg: BitSharesOperation) -> None:
+        self.data = msg
+
+        self.nodes = OrderedDict(
+            [
+                ("operation_id", OperationId),
+                ("fee", Asset),
+                # ("operation_name", OperationCls) is added in parse
+            ]
+        )
+
+    def parse(self) -> "Operation":
+        if self.data.operation_name is None:
+            raise ValueError("operation_name")
+
+        self.op_name = self.data.operation_name
+
+        # add the operation_id to the msg such that Node.parse() can be applied
+        op_id = OperationTable.get_id(self.op_name)
+        setattr(self.data, "operation_id", op_id)
+
+        # add the operation itself
+        OperationCls = OperationTable.get_cls(self.op_name)
+        self.nodes[self.op_name] = OperationCls
+
+        return super().parse()
+
+    def listify(self) -> (str, List[str]):
+        return (self.op_name, self[self.op_name].listify())
+
+
+class TransferOperation(Node):
+    def __init__(self, msg: BitSharesTransferOperation) -> None:
+        self.data = msg
+
+        self.nodes = OrderedDict(
+            [
+                ("sender", ObjectId),
+                ("receiver", ObjectId),
+                ("amount", Asset),
+                ("memo", Optional(Memo)),
+                ("extensions", EmptyExtension),
+            ]
+        )
+
+    def listify(self) -> List[str]:
+        return [
+            "from:",
+            str(self["sender"]),
+            "to:",
+            str(self["receiver"]),
+            "amount:",
+            str(self["amount"]["amount"]),
```

Comment:
Awkward enough, the amount in the transaction is in satoshi .. 
Let's leave it that way right now. Otherwise we'd need to hand over the precision via USB and trust the computer to provide the proper precision.

Could hard code the precisions .. but really have no other secure way of dealing with this.

Reply:
EOS used the `asset_id` to pass the precision. They replaced the `space_id` with precision for example. Or I add an optional field to the `BitSharesAsset` protob message.